File: Algorithms/Mathematical Algorithms/PowerOfANumberIterative.py
__author__ = 'Kalpak Seal'

def calculate_pow(x, y):
    res = 1.0
    exp = y
    base = x

    if y < 0:
        base = 1 / x
        exp *= -1

    while exp > 0:
        if exp & 1:
            res *= base

        exp = exp >> 1
        base *= base

    return res

if __name__ == '__main__':
    print(calculate_pow(4, 3))
    print(calculate_pow(4, -3))

File: Algorithms/Mathematical Algorithms/PowerofANumberIterative.c
#include <stdio.h>

/* Iterative Function to calculate (x^y) in O(log y) */
float power(int x, int y)
{
	float res = 1.0;	 // Initialize result
	float base = (float)x;
	int exp = y;
	
	if (y < 0)
	{
	    base = 1/(float)x;
	    exp = y * (-1);
	}
    
	while (exp > 0)
	{
		// If exp is odd, multiply base with result
		if (exp & 1)
		{
			res = res*base;
		}

		// exp must be even now
		exp = exp>>1; 
		base *= base; // Change base to base^2
	}
	return res;
}

int main(void){
    printf("%f \n", power(2, 2));
    return 0;
}
